File: C++/768.cpp
__________________________________________________________________________________________________
sample 12 ms submission
class Solution {
public:
    int maxChunksToSorted(vector<int>& arr) {
        vector<int> mins(arr.size(), INT_MAX);
        for (int i=arr.size()-2;i>=0;i--)
            mins[i] = min(mins[i+1], arr[i+1]);
        int maxV = -1, count = 0;
        for (int i=0; i<arr.size();i++) {
            maxV = max(maxV, arr[i]);
            if (maxV<=mins[i]) count += 1;
        }
        return count;
    }
};
__________________________________________________________________________________________________
sample 9800 kb submission
class Solution {
public:
    int maxChunksToSorted(vector<int>& arr) {
        int N = arr.size();
        
        vector<int> aux(N,arr.back());
        for(int i=N-2;i>=0;i--)
            aux[i] = min(arr[i],aux[i+1]);
        
        int chunk = 0, max_v = -1;
        for(int i=0;i<N;i++){
            max_v = max_v == -1 ? arr[i] : max(max_v,arr[i]);
            if(i == N-1 || max_v <= aux[i+1]){
                chunk++;
                max_v = -1;
            }
        }
        return  chunk;
    }
};
__________________________________________________________________________________________________

File: Python3/768.py
__________________________________________________________________________________________________
sample 76 ms submission
class Solution:
    def maxChunksToSorted(self, arr: List[int]) -> int:
        stack = []
        
        for n in arr:
            if not stack or stack[-1] <= n:
                stack.append(n)
            else:
                tmp = stack[-1]
                while stack and stack[-1] > n:
                    stack.pop()
                stack.append(tmp)
        
        return len(stack)
__________________________________________________________________________________________________
sample 80 ms submission
class Solution:
    def maxChunksToSorted(self, arr: List[int]) -> int:
        sorted_arr = sorted(arr)
        D = {}
        for i in range(len(sorted_arr)):
            if sorted_arr[i] not in D:
                D[sorted_arr[i]] = i
        earliest_finish_index = 0
        count = 0
        res = 0
        for i in range(len(arr)):
            if D[arr[i]] > earliest_finish_index:
                earliest_finish_index = D[arr[i]]
                count = 1
            elif D[arr[i]] == earliest_finish_index:
                count += 1
            if earliest_finish_index <= i and i + 1 - count >= earliest_finish_index:
                res += 1
        return res
__________________________________________________________________________________________________
